dotnet/src/SemanticKernel.Abstractions/Kernel.cs

```diff
@@ -191,26 +202,52 @@ public CultureInfo Culture
         this._data;
 
     #region internal ===============================================================================
-    internal bool OnFunctionInvoking(FunctionInvokingEventArgs eventArgs)
+    internal FunctionInvokingEventArgs? OnFunctionInvoking(KernelFunction function, SKContext context)
+    {
+        if (this.FunctionInvoking is null)
+        {
+            return null;
+        }
+
+        var eventArgs = new FunctionInvokingEventArgs(function, context);
+        this.FunctionInvoking.Invoke(this, eventArgs);
+        return eventArgs;
+    }
+
+    internal FunctionInvokedEventArgs? OnFunctionInvoked(KernelFunction function, FunctionResult result)
+    {
+        if (this.FunctionInvoked is null)
```

Comment:
There's a race condition here: if code concurrently unsubscribes their handler and it's the last one, the invocation could null ref, as you'll check and see it's not null but then it'll become null after you check. That's handled for you with the ?. syntax, as it makes a local copy before the null check. Wanting to return the event args complicates things, though. So you'll want to store it into a local first and then do all operations (null check and invocation) with that local.